api_docs/updating-group-settings.md

```diff
@@ -0,0 +1,32 @@
+# Updating group settings
+
```

Comment:
We should probably add an introductory line here like "Client should pass a group setting update object", but am not sure how to phrase it.

Reply:
Yeah I'll do an introduction revision pass on this before merging, looks like a great start though.

Reply:
Ended up doing quite a bit of reworking to make this nice:

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 6c7401c8b0..a54dd829cf 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -23,8 +23,8 @@ format used by the Zulip server that they are interacting with.
 **Feature level 260**:
 
 * [`PATCH /user_groups/{user_group_id}`](/api/update-user-group):
-  `can_mention_group` parameter now also accepts object value
-  containing `old` and `new` fields.
+  Updating `can_mention_group` now uses a race-resistant format where
+  the client sends the expected `old` value and desired `new` value.
 
 **Feature level 259**:
 
diff --git a/api_docs/group-setting-values.md b/api_docs/group-setting-values.md
new file mode 100644
index 0000000000..c6a9f9d8b8
--- /dev/null
+++ b/api_docs/group-setting-values.md
@@ -0,0 +1,79 @@
+# Group-setting values
+
+Settings defining permissions in Zulip are increasingly represented
+using [user groups](/help/user-groups), which offer much more flexible
+configuration than the older [roles](/api/roles-and-permissions) system.
+
+In the API, these settings are represented using a **group-setting
+value**, which can take two forms:
+
+- An integer user group ID, which can be either a named user group
+  visible in the UI or a [role-based system group](#system-groups).
+- An object with fields `direct_member_ids` containing a list of
+  integer user IDs and `direct_subgroup_ids` containing a list of
+  integer group IDs. The setting's value is the union of the
+  identified collection of users and groups.
+
+Group-setting values in the object form function very much like a
+formal user group object, without requiring the naming and UI clutter
+overhead involved with creating a visible user group just to store the
+value of a single setting.
+
+The server will canonicalize an object with empty `direct_member_ids`
+and with `direct_subgroup_ids` containing just the given group ID to
+the integer format.
+
+## System groups
+
+The Zulip server maintains a collection of system groups that
+correspond to the users with a given role; this makes it convenient to
+store concepts like "all administrators" in a group-setting
+value. These use a special naming convention and can be recognized by
+the `is_system_group` property on their group object.
+
+The following system groups are maintained by the Zulip server:
+
+- `role:internet`: Everyone on the Internet has this permission; this
+  is used toconfigure the [public access
+  option](/help/public-access-option).
+- `role:everyone`: All users, including guests, are in this group.
+- `role:members`: All members of the organization.
+- `role:fullmembers`: All [full
+  members](https://zulip.com/api/roles-and-permissions#determining-if-a-user-is-a-full-member)
+  of the organization are in this group.
+- `role:moderators`: All users with at least the moderator role.
+- `role:administrators`: All users with at least the administrator
+  role.
+- `role:owners`: All users with the owner role.
+- `role:nobody`: The formal empty group. Used to represent disabling a
+  feature, by giving nobody permission to use it.
+
+Client UI for setting a permission can display system groups as
+special/simpler options.
+
+System groups should generally not be displayed in UI for
+administering an organization's user groups, since they are not
+directly mutable.
+
+## Updating group-setting values
+
+A **group-setting update** is an object with a `new` field and an
+optional `old` field, each containing a group-setting value. The
+setting's value will be set to the membership expressed by the `new`
+field.
+
+The `old` field expresses the client's understanding of the current
+value of the setting. If the `old` field is present and does not match
+the actual current value of the setting, then the request will fail
+with error code `EXPECTATION_MISMATCH` and no changes will be applied.
+
+When a user edits the setting in a UI, the resulting API request
+should generally always include the `old` field, giving the value
+the list had when the user started editing. This accurately expresses
+the user's intent, and if two users edit the same list around the
+same time, it prevents a situation where the second change
+accidentally reverts the first one without either user noticing.
+
+Omitting `old` is appropriate where the intent really is a new complete
+list rather than an edit, for example in an integration that syncs the
+list from an external source of truth.
diff --git a/api_docs/roles-and-permissions.md b/api_docs/roles-and-permissions.md
index 662ae6513d..db3c4acba9 100644
--- a/api_docs/roles-and-permissions.md
+++ b/api_docs/roles-and-permissions.md
@@ -102,6 +102,11 @@ and owners.
 Note that specific settings and policies in the Zulip API that use these
 permission levels will likely support a subset of those listed above.
 
+## Group-based permissions
+
+Some settings have been migrated to a more flexible system based on
+[user groups](/api/group-setting-values).
+
 ## Determining if a user is a full member
 
 When a Zulip organization has set up a [waiting period before new members
diff --git a/api_docs/sidebar_index.md b/api_docs/sidebar_index.md
index 099164b9c5..10540e7946 100644
--- a/api_docs/sidebar_index.md
+++ b/api_docs/sidebar_index.md
@@ -21,7 +21,7 @@
 * [HTTP headers](/api/http-headers)
 * [Error handling](/api/rest-error-handling)
 * [Roles and permissions](/api/roles-and-permissions)
-* [Updating group settings](/api/updating-group-settings)
+* [Group-setting values](/api/group-setting-values)
 * [Message formatting](/api/message-formatting)
 * [Client libraries](/api/client-libraries)
 * [API changelog](/api/changelog)
diff --git a/api_docs/updating-group-settings.md b/api_docs/updating-group-settings.md
deleted file mode 100644
index c819756650..0000000000
--- a/api_docs/updating-group-settings.md
+++ /dev/null
@@ -1,32 +0,0 @@
-# Updating group settings
-
-A **group-setting update** is an object with a `new` field and
-an optional `old` field, each containing a group-setting value
-(as defined below). The setting's value will be set to the
-membership expressed by the `new` field.
-
-The `old` field expresses the expected current value of the setting.
-If the `old` field is present and does not match the actual current
-value of the setting, then the request will fail with error code
-`EXPECTATION_MISMATCH` and no changes will be applied.
-
-When a user edits the setting in a UI, the resulting API request
-should generally always include the `old` field, giving the value
-the list had when the user started editing. This accurately expresses
-the user's intent, and if two users edit the same list around the
-same time, it prevents a situation where the second change
-accidentally reverts the first one without either user noticing.
-
-Omitting `old` is appropriate where the intent really is a new complete
-list rather than an edit, for example in an integration that syncs the
-list from an external source of truth.
-
-A **group-setting value** can take two forms: either an integer group ID,
-or an object with fields `direct_member_ids` containing a list of integer
-user IDs and `direct_subgroup_ids` containing a list of integer group IDs.
-
-In the object form, the setting's value consists of the identified users
-and groups.
-
-The integer form is equivalent to an object with empty `direct_member_ids`
-and with `direct_subgroup_ids` containing just the given group ID.
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index cb0fe8e75c..cedfe2537e 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -3167,7 +3167,7 @@ paths:
                                         changed.
                                     can_mention_group:
                                       allOf:
-                                        - $ref: "#/components/schemas/CanMentionGroup"
+                                        - $ref: "#/components/schemas/GroupSettingValue"
                                         - description: |
                                             Either the ID of a named user group that has permission
                                             to mention the group, or an object describing the set of
@@ -18589,9 +18589,8 @@ paths:
                 can_mention_group:
                   allOf:
                     - description: |
-                        Either the ID of a named user group that has permission to
-                        mention the group, or an object describing the set of users
-                        and groups who have permission mention the new group.
+                        A [group-setting value](/api/group-setting-values) defining the set
+                        of users who have permission to mention the new group.
 
                         This setting cannot be set to `"role:internet"` and
                         `"role:owners"` system groups.
@@ -18604,7 +18603,7 @@ paths:
 
                         New in Zulip 8.0 (feature level 191). Previously, groups
                         could be mentioned if and only if they were not system groups.
-                    - $ref: "#/components/schemas/CanMentionGroup"
+                    - $ref: "#/components/schemas/GroupSettingValue"
                   example: 11
               required:
                 - name
@@ -18743,10 +18742,9 @@ paths:
                   example: The marketing team.
                 can_mention_group:
                   description: |
-                    The set of users who have permission to mention this group.
-
-                    This parameter is expressed as a [group-setting update](/api/updating-group-settings).
-                    See [details](/api/updating-group-settings) on that format.
+                    The set of users who have permission to mention
+                    this group, expressed as an [update to a
+                    group-setting value](/api/group-valued-settings#updating-group-setting-values).
 
                     This setting cannot be set to `"role:internet"` and `"role:owners"`
                     system groups.
@@ -18768,15 +18766,15 @@ paths:
                     new:
                       allOf:
                         - description: |
-                            The new value for who would have the permission to
-                            mention the group.
-                        - $ref: "#/components/schemas/CanMentionGroup"
+                            The new [group-setting value](/api/group-setting-values) for who would
+                            have the permission to mention the group.
+                        - $ref: "#/components/schemas/GroupSettingValue"
                     old:
                       allOf:
                         - description: |
-                            The expected current value for who has the permission
-                            to mention the group.
-                        - $ref: "#/components/schemas/CanMentionGroup"
+                            The expected current [group-setting value](/api/group-setting-values)
+                            for who has the permission to mention the group.
+                        - $ref: "#/components/schemas/GroupSettingValue"
                   required:
                     - new
                   example:
@@ -18904,11 +18902,10 @@ paths:
                                 **Changes**: New in Zulip 5.0 (feature level 93).
                             can_mention_group:
                               allOf:
-                                - $ref: "#/components/schemas/CanMentionGroup"
+                                - $ref: "#/components/schemas/GroupSettingValue"
                                 - description: |
-                                    Either the ID of a named user group that has permission to
-                                    mention the group, or an object describing the set of users
-                                    and groups who have permission to mention the group.
+                                    A [group-setting value](/api/group-setting-values) defining the set
+                                    of users who have permission to mention the new group.
 
                                     **Changes**: Before Zulip 9.0 (feature level 258), the
                                     `can_mention_group` field was always an integer.
@@ -20088,11 +20085,10 @@ components:
             **Changes**: New in Zulip 5.0 (feature level 93).
         can_mention_group:
           allOf:
-            - $ref: "#/components/schemas/CanMentionGroup"
+            - $ref: "#/components/schemas/GroupSettingValue"
             - description: |
-                Either the ID of a named user group that has permission to
-                mention the group, or an object describing the set of users
-                and groups who have permission mention the group.
+                A [group-setting value](/api/group-setting-values) defining the set
+                of users who have permission to mention the new group.
 
                 **Changes**: Before Zulip 9.0 (feature level 258), the
                 `can_mention_group` field was always an integer.
@@ -20102,26 +20098,26 @@ components:
 
                 New in Zulip 8.0 (feature level 191). Previously, groups
                 could be mentioned if and only if they were not system groups.
-    CanMentionGroup:
+    GroupSettingValue:
       oneOf:
         - type: object
           additionalProperties: false
           properties:
             direct_members:
               description: |
-                The list of user IDs that have permission to
-                mention the group.
+                The list of IDs of individual users in the collection of users with this permission.
               type: array
               items:
                 type: integer
             direct_subgroups:
               description: |
-                The list of user group IDs that have permission
-                to mention the group.
+                The list of IDs of the groups in the collection of users with this permission.
               type: array
               items:
                 type: integer
         - type: integer
+          description: |
+            The ID of a [user group](/help/user-groups) with this permission.
     Invite:
       type: object
       description: |
```